### DISCUSSION TITLE
Question about (subgraph)-isomorphism and -(monomorphism)

### DISCUSSION THREAD

**Discussion by Nado15** [networkx/networkx#7615] [Q&A] [answered]
Hi, 

First of all, thank you very much for the work you are doing here. I am totally satisfied with the library. I came across a case that I would like to discuss here. 

For an algorithm of mine, I have to solve the subgraph counting problem (find all occurrences of graph A in graph B). For this I used the function `subgraph_isomorphisms_iter()`. However, I came across the following graph for which I cannot find a solution:

![Graph](https://github.com/user-attachments/assets/4e211126-dbcf-42fa-a8c9-7e5c8ef1e537)

I had expected to find all 4(!) subgraphs. However, the search does not return a single mapping. 

So after doing more research, I stumbled upon `subgraph_monomorphisms_iter()`, however the search for `DiGraphMatcher(A, B).subgraph_monomorphisms_iter()` does not return a mapping either. However, if I swap the order of the argument `DiGraphMatcher(B, A).subgraph_monomorphisms_iter()` I find all 4 subgraphs. 

Maybe someone can enlighten me, as I am quite confused.

**Comment by dschult** [comment, marked as answer]
The order of arguments in subgraph isomorphism is important and the language surrounding it makes switching the order too easy to happen. The isomorphism docs have [a section on subgraph isomorphism](https://networkx.org/documentation/stable/reference/algorithms/isomorphism.vf2.html#subgraph-isomorphism) that attempts to make this more clear. 

> In the VF2 literature, a mapping M is said to be a graph-subgraph isomorphism iff M is an isomorphism between G2 and a subgraph of G1. Thus, to say that G1 and G2 are graph-subgraph isomorphic is to say that a subgraph of G1 is isomorphic to G2.

> Other literature uses the phrase ‘subgraph isomorphic’ as in ‘G1 does not have a subgraph isomorphic to G2’. Another use is as an in adverb for isomorphic. Thus, to say that G1 and G2 are subgraph isomorphic is to say that a subgraph of G1 is isomorphic to G2.

The main idea is that `GraphMatcher(G, H)` is looking for a mapping between H and a subgraph of G.  G is the (potentially) bigger of the two. H is the subgraph we're looking for.  The paper literature seems to use language with either direction depending on the author's preference, and we picked this order for inputs G and H.

Does this help?  Are there improvements we could make to the docs?

**Reply by Nado15** [reply]
Hi, thank you very much for your answer. Yes that answers my question but I have another one.

I don't quite understand why I can't find any solutions with `...(B,A).subgraph_isomorphisms_iter()` but I can find all 4 possible subgraphs with `...(B,A).subgraph_monomorphisms_iter()`. 

Just for testing I used [graph-tool](https://graph-tool.skewed.de/) with `subgraph_ismorphism()` and found all 4 possible subgraphs.

I looked up the [documentation](https://networkx.org/documentation/stable/reference/algorithms/isomorphism.vf2.html#subgraph-isomorphism) and found the following part:

> ...Edge-induced subgraph isomorphisms are not directly supported...

Is this possibly the reason?

Graph A is the one on the left side
Graph B is the one on the right side

**Reply by dschult** [reply]
There are varying definitions of "subgraph - isomorphism". In [the NetworkX docs on subgraph isomorphism](https://networkx.org/documentation/stable/reference/algorithms/isomorphism.vf2.html#subgraph-isomorphism) we state that we use the term "subgraph - isomorphism" to mean what some would call a "node induced subgraph isomorphism". That is, the edges must contain all and only the edges connecting the nodes in the subgraph. A "monomorphism" means that the mapping need not maintain the edges between nodes in the subgraph. That is, some extra edges can appear in the bigger graph.

I don't think edge-induced subgraphs are considered by any graph packages. The usual difference is that some use "subgraph" to be a node induced subgraph, while others use "subgraph" to mean that the node-set and edge-set are subsets of the original. We call that a subgraph monomorphism because it doesn't preserve the edge structure on the nodes (so it's not "iso").
